### ISSUE TITLE
Admin bar: Cannot disable components.

### ISSUE THREAD

**Issue by bd0bd** [backdrop/backdrop-issues#2556]
![bd](https://cloud.githubusercontent.com/assets/13515302/23159439/d059266e-f833-11e6-84fe-a8ade2caaaea.png)

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/1797 (changes the admin_bar.settings.json data structure)
PR by @quicksketch: https://github.com/backdrop/backdrop/pull/1808 (changes the way the form saves to use the previous structure)

**Comment by klonos** [comment]
This is a bug. I've figured a couple of was to fix this. Still looking into which way would be the best, but it boils down to this:

- When components are disabled, this is what gets saved in `admin_bar.settings.json`:

```json
{
    "_config_name": "admin_bar.settings",
    "margin_top": 1,
    "position_fixed": 0,
    "components": {
        "admin_bar.icon": 0,
        "admin_bar.menu": 0,
        "admin_bar.search": 0,
        "admin_bar.users": 0,
        "admin_bar.account": 0
    }
}
```

- When components are enabled, this is what gets saved in `admin_bar.settings.json`:

```json
{
    "_config_name": "admin_bar.settings",
    "margin_top": 1,
    "position_fixed": 0,
    "components": {
        "admin_bar.icon": "admin_bar.icon",
        "admin_bar.menu": "admin_bar.menu",
        "admin_bar.search": "admin_bar.search",
        "admin_bar.users": "admin_bar.users",
        "admin_bar.account": "admin_bar.account"
    }
}
```

I would expect those values to be `1` instead.

In [`admin_bar_output()`](https://github.com/backdrop/backdrop/blob/1.x/core/modules/admin_bar/admin_bar.inc#L15), we are [loading the config](https://github.com/backdrop/backdrop/blob/1.x/core/modules/admin_bar/admin_bar.inc#L17) and [storing the components values in a `$components` array](https://github.com/backdrop/backdrop/blob/1.x/core/modules/admin_bar/admin_bar.inc#L36):

```php
$config = config('admin_bar.settings');
...
$components = $config->get('components');`
```

Later,[ when checking whether a certain component is enabled or not](https://github.com/backdrop/backdrop/blob/1.x/core/modules/admin_bar/admin_bar.inc#L71), we do:

```php
if (in_array('admin_bar.icon', $components) || $complete) {
```

...which will always return true (and that's why all components are shown all the time no matter whether they've been set to disabled or not). What we could be doing instead is:

```php
if ($components['admin_bar.menu'] === 'admin_bar.menu' || $complete) {
```

...but that looks messy. We could instead be doing this with the `$components` array:

```php
    foreach ($components as $component => $component_value) {
      if ($component_value != '') {
        $components[$component] = 1;
      }
      else {
        $components[$component] = 0;
      }
    }
```

...and then:

```php
if ($components['admin_bar.menu'] || $complete) {
```

...which works (and looks much better), but why not save the values of the components properly (`0` or `1`) in the config file in the first place?

Anyways, I'm on to this and will file a PR soon.

**Comment by serundeputy** [comment]
@klonos 

> ...which works (and looks much better), but why not save the values of the components properly (0 or 1) in the config file in the first place?

yes, I would say save them as `0` or `1` in the config and focus your PR on that method.

**Comment by klonos** [comment]
Yes, but (unless I'm horribly mistaken) that's an issue with the Form API and how multiple checkbox values are saved in config by default (without using a form `*_submit` handler function). This should be fixed in core (#2557), and then this issue here would simply "fix itself".

**Comment by klonos** [comment]
I have created a simple PR for review that just solves the immediate issue of the the settings not being applied. This keeps the old way of saving things in the config.

I have also started working on an alternative approach (using a form submit handler). So far I have gotten the settings to save properly in the config with `1`/`0` and the default values to be picked up in the form. I have changed the function name from `admin_bar_theme_settings()` to `admin_bar_theme_settings_form` though. Is that allowed within the 1.x cycle? I also had to rename the settings/options from `admin_bar.xyz` to just `xyz` (since that dot caused the creation of an extraneous `admin_bar` child entry in the json file), but now I've broken the rendering of the admin bar (every component gets rendered no matter if disabled/enabled) ...which defeats the purpose.

On top of everything, there's some fancy use of caching going on in that form I think, since any `dpm` is rendered only after cache clear and then page refresh. This makes it really hard for a beginner coder like me to troubleshoot further. Anyways, it's getting late (or early, depending on what one calls 6am) and I'm getting cranky, so I guess I'm gonna have to pick this up again tomorrow ...after I had enough 💤

**Comment by klonos** [comment]
...note to self: remember to also change the default config .json accordingly.

**Comment by olafgrabienski** [comment]
@klonos - I tested the "simple PR". Could disabe and reenable all Admin bar components.

**Comment by klonos** [comment]
Thanx @olafgrabienski simple works, but it leaves things in a messy situation IMHO both code-wise and configuration-wise. Unfortunately, I have very little time lately, so things need to wait. What I might do tonight is post a separate PR with whatever I have as the alternative, "proper" approach and leave it up for review (even though not 100% working).

**Comment by quicksketch** [comment]
Thanks @klonos for the PR! I'm concerned that we're changing the data structure of the admin_bar.settings.json form. That in itself may be a minor API break, and would require an update hook to flip between data formats (to do it properly). An alternative is to keep the current data structure and code, but just change the way the form submits to save in the expected format. This would essentially fix a regression created by #675.

PR https://github.com/backdrop/backdrop/pull/1808.

**Comment by klonos** [comment]
@quicksketch with https://github.com/backdrop/backdrop/pull/1808 you'd get:

```json
{
    "_config_name": "admin_bar.settings",
    "margin_top": 1,
    "position_fixed": 1,
    "components": [
        "admin_bar.icon",
        "admin_bar.menu",
        "admin_bar.search",
    ]
}
```

...which is OK if you want to check which components are enabled, but in order to check which ones are disabled, you'd need to know their "machine names" and check whether they are missing from the list. Which is not as easy as if you have them all on the list and specifically checked for enabled/disabled:

```json
{
    "_config_name": "admin_bar.settings",
    "margin_top": 1,
    "position_fixed": 1,
    "components": {
        "admin_bar.icon": 1,
        "admin_bar.menu": 1,
        "admin_bar.search": 1,
        "admin_bar.users": 0,
        "admin_bar.account": 0
    }
}
```

Can you please take a look at my new PR: https://github.com/backdrop/backdrop/pull/1811

**Comment by klonos** [comment]
...basically what I was trying to say is that with the first example config in my comment above, you cannot do a `foreach` after a `$config->get('components')`.

**Comment by klonos** [comment]
...or more precisely, that the `foreach` would only loop through enabled components and not through all possible ones.

**Comment by serundeputy** [comment]
Thanks @bd0bd for the report, @klonos for the leg work, and @quicksketch for iterating on it.  Teamwork! 

https://github.com/backdrop/backdrop/pull/1808 fixes the immediate bug and works well.

Merged into `1.x` and `1.6.x` will ship with `1.6.4`
